lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -124,11 +120,27 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
     -- then you can skip requesting the headers and go straight to requesting
     -- the block itself.
     blocks <- if singleBlockHeader
-              then requestBlocks (NewestFirst (pure tipHeader))
-              else requestHeaders >>= requestBlocks
+              then requestBlocks (OldestFirst (one tipHeader))
+              else twoPhaseAction
     pure (OldestFirst (reverse (toList blocks)))
   where
 
+    twoPhaseAction = do
+        headers <- toOldestFirst <$> requestHeaders
+        getLcaMainChain logic headers >>= \case
```

Comment:
How much work would it be to make something like `getLcaMainChain` that, instead of giving the lca hash, gives the suffix that we compute after this? So that we wouldn't have to `dropUntilLca` again? This could save a traversal and eliminate that error case on line 140.

Reply:
`dropUntilLca` usually does exactly 0 steps, because usually headers list is already `[c1..cn]`, where `c1` is a child of our tip `c0`, so it's a direct continuation. I don't see any need to optimize here, it's a _really_ rare case that this `dropUntilLca` will actually do something, because rollbacks are pretty rare. And if they are, it's anyway super quick and happens only once (there can't be more than one serious rollback).